mlflow/models/signature.py

```diff
@@ -340,32 +346,71 @@ def _extract_type_hints(f, input_arg_index):
     return _TypeHints(hints.get(arg_name), hints.get("return"))
 
 
-def _infer_signature_from_type_hints(func, input_arg_index, input_example=None):
-    hints = _extract_type_hints(func, input_arg_index)
-    if hints.input is None:
+def _is_context_in_predict_function_signature(*, func=None, parameters=None):
+    if parameters is None:
+        if func is None:
+            raise ValueError("Either `func` or `parameters` must be provided.")
+        parameters = inspect.signature(func).parameters
+    return (
+        # predict(self, context, model_input, ...)
+        "context" in parameters
+        # predict(self, ctx, model_input, ...) ctx can be any parameter name
+        or len([param for param in parameters if param != "params"]) == 2
+    )
+
+
+def _infer_signature_from_type_hints(func, type_hints: _TypeHints, input_example=None):
+    if type_hints.input is None:
         return None
 
     params = None
     params_key = "params"
     if _contains_params(input_example):
         input_example, params = input_example
 
-    input_schema = _infer_schema_from_type_hint(hints.input, input_example) if hints.input else None
+    try:
+        input_schema = _infer_schema_from_type_hint(type_hints.input)
+    except InvalidTypeHintException as e:
+        warnings.warn(e.message, stacklevel=2)
+        return None
     params_schema = _infer_param_schema(params) if params else None
-    input_arg_name = _get_arg_names(func)[input_arg_index]
-    if input_example:
-        inputs = {input_arg_name: input_example}
+    if input_schema and input_example:
+        try:
+            _validate_example_against_type_hint(example=input_example, type_hint=type_hints.input)
+        except Exception as e:
+            raise MlflowException.invalid_parameter_value(
+                "Input example is not compatible with the type hint of the `predict` function. "
+                f"Type hint: {type_hints.input}."
+            ) from e
+        kwargs = {}
         if params and params_key in inspect.signature(func).parameters:
-            inputs[params_key] = params
+            kwargs[params_key] = params
         # This is for PythonModel's predict function
-        if input_arg_index == 1:
-            inputs["context"] = None
-        output_example = func(**inputs)
+        if _is_context_in_predict_function_signature(func=func):
+            inputs = [None, input_example]
+        else:
+            inputs = [input_example]
+        # Note: This has the assumption that input data is not converted at all
+        # TODO: should we catch exception here?
+        output_example = func(*inputs, **kwargs)
```

Comment:
1. Can we add a logger info message like "Running model prediction to validate output schema"? We heard some users get confused by long running logging operation.
2. I think we should catch an exception here. Generally we don't throw for incorrect signature and input example, so it's more consistent not to throw here as well.

Reply:
Make sense :)